doc/dotNET SDK Proposal.md

```diff
@@ -0,0 +1,232 @@
+
+**Build on top of "Core".**
+We will build the .NET Host SDK on top of the [SDK Core library](https://github.com/temporalio/sdk-core) (aka "Core"). Core is a native library written in Rust that provides common abstractions for Temporal state machines, for remote GRPC invocations, and for other aspects required to build runtime-specific SDKs with consistent functionality. Currently, Core is used as the basis for the Typescript/Javascript SDK.
+
+Where Core features are lacking, we will make case-by-case decisions to add required functionality to Core or to implement it directly in .NET.
+
+**SDK Engine hosts workflows and activities and manages their execution.**
+The center of the .NET SDK is the _Engine_. The Engine keeps a list of workflows and activities hosted in a particular worker host, and a mapping of active workflow (and activity) instances and their respective local execution states. The Engine contains a message loop that polls Core for work items, such as workflow activations, activity tasks, etc. Based on that, the engine invokes and manages user-provided workflow / activity implementations.
+
+.NET-based Temporal _activities_ are - similar to other languages supported by Temporal -  (almost) arbitrary components. There is only a small amount of runtime management involved into managing the execution of activities. In this high-level overview we therefore focus on the execution of .NET-based Temporal _workflows_.
+
+### Workflow Restrictions
+
+In general, .NET-based workflows provide very similar guarantees and follow very similar restrictions as other languages. A key aspect of the restrictions across all Temporal languages is that workflows must be _deterministic_. This requires special rules around any APIs that may not produce identical results if run multiple times with the same input. Most prominently this includes APIs that interact with the environment (e.g. I/O), and threading/parallelism related APIs (because most common multi-threading implementations are not deterministic in nature). Here we focus on .NET-specific nuances.
+
+**Workflows are strictly single-threaded and asynchronous:**
+
+**In the first version of the .NET SDK, workflows must be <u>strictly logically single-threaded</u>.**
+It is not permitted to use any APIs that create or manipulate threads (implicitly or explicitly), execute thread-synchronization primitives, or perform any kind of parallel execution. Note that being _logically_ single-threaded does not mean that a given workflow is bound to a specific thread. On contrary, the Engine will suspend workflows that block waiting for Temporal events to occur, and it will resume such workflows on any (likely a different) thread when those events eventually occur. However, _logically_ workflow must be strictly single-threaded. I.e., while the execution may be performed on different underlying threads, it must never _fork_ into parallel (aka concurrent) execution sequences.
```

Comment:
> I think concurrency (what you call asynchrony) is expected in Workflows, we should definitely prohibit true parallelism.

To make sure we use the same terms:

* "Asynchrony" is about not worrying about when an operation executes or completes and only carrying about its completion / result.
* "Parallelism" is about executing something in parallel to each other (i.e. at the same time). Strictly speaking, parallel operations are a subset of concurrent operations.
* "Concurrency" is about running things at the same time. Strictly speaking, concurrent operations on a single core machine would not be parallel. Asynchronous operations may (or not be concurrent), but in practice they often (usually) are.

In .NET. Taks were first used to address scenarios from Parallelism (in .NET 4.0), and then from Synchrony (in .NET 4.5), which makes them confusing. :)

I'll clarify this in text and will check that the terms are used consistently.

In the SDK we want asynchrony, but no parallelism.
There is a section in the docs that says that in the first version we want to avoid tackling the challenge of taking parallel code and turning it into concurrent, non-parallel code by creating a custom scheduler. If there is demand for it, we can and will tackle is subsequently. 

Does this clarify things?

Reply:
Additionally
>parallel (aka concurrent)

These terms do have different and well-defined meanings. Parallelism is when two things actually execute simultaneously. Concurrency is the *appearance* of simultaneous execution. A single-core processor can execute programs concurrently, because the OS forces them to yield to each other, but can never execute them in parallel.

In this context, it means workflows can perform concurrent, but never parallel, actions (which we will enforce for/require of the user)

Reply:
Yeah, I think we are saying the same thing with slightly different words. :)

Reply:
✔️